zqd/space/space.go

```diff
@@ -32,232 +29,180 @@ var (
 	ErrSpaceNotExist       = zqe.E(zqe.NotFound, "space does not exist")
 )
 
+type Space interface {
+	ID() api.SpaceID
+	Storage() storage.Storage
+	Info(context.Context) (api.SpaceInfo, error)
+
+	// StartOp is called to register an operation is in progress; the
+	// returned cancel function must be called when the operation is done.
+	StartOp(context.Context) (context.Context, context.CancelFunc, error)
+
+	// Delete cancels any outstanding operations, then removes the space's path
+	// and data dir (should the data dir be different then the space's path).
+	// Intended to be called from Manager.Delete().
+	delete() error
+
+	Update(api.SpacePutRequest) error
+}
+
 func newSpaceID() api.SpaceID {
 	id := ksuid.New()
 	return api.SpaceID(fmt.Sprintf("sp_%s", id.String()))
 }
 
-type Space struct {
-	Storage storage.Storage
-
-	path string
-	conf config
-
+type guard struct {
 	// state about operations in progress
 	opMutex       sync.Mutex
-	active        int
 	deletePending bool
 
 	wg sync.WaitGroup
 	// closed to signal non-delete ops should terminate
 	cancelChan chan struct{}
 }
 
-func loadSpace(path string, conf config) (*Space, error) {
-	datapath := conf.DataPath
-	if datapath == "." {
-		datapath = path
-	}
-	s, err := loader.Load(datapath)
-	if err != nil {
-		return nil, err
+func newGuard() *guard {
+	return &guard{
+		cancelChan: make(chan struct{}),
 	}
-	return &Space{
-		Storage:    s,
-		path:       path,
-		conf:       conf,
-		cancelChan: make(chan struct{}, 0),
-	}, nil
 }
 
-// StartSpaceOp registers that an operation on this space is in progress.
-// If the space is pending deletion, an error is returned.
-// Otherwise, this returns a new context, and a done function that must
-// be called when the operation completes.
-func (s *Space) StartSpaceOp(ctx context.Context) (context.Context, context.CancelFunc, error) {
-	s.opMutex.Lock()
-	defer s.opMutex.Unlock()
+func (g *guard) acquire(ctx context.Context) (context.Context, context.CancelFunc, error) {
+	g.opMutex.Lock()
+	defer g.opMutex.Unlock()
 
-	if s.deletePending {
+	if g.deletePending {
 		return ctx, func() {}, zqe.E(zqe.Conflict, "space is pending deletion")
 	}
 
-	s.wg.Add(1)
+	g.wg.Add(1)
 
 	ctx, cancel := context.WithCancel(ctx)
 	go func() {
 		select {
 		case <-ctx.Done():
-		case <-s.cancelChan:
+		case <-g.cancelChan:
 			cancel()
 		}
 	}()
 
 	done := func() {
-		s.opMutex.Lock()
-		defer s.opMutex.Unlock()
+		g.opMutex.Lock()
+		defer g.opMutex.Unlock()
```

Comment:
I don't believe this mutex is necessary.

Reply:
thanks, I believe this was left over from when there was an 'active' integer counter to track the number of outstanding operations, will remove.